File: CS/MeffedNavigationService/ViewModel/MenuViewModel.cs
using DevExpress.Xpf.Mvvm;
using DevExpress.Xpf.Mvvm.UI;
using DevExpress.Xpf.WindowsUI.Navigation;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel.Composition;
using System.ComponentModel.Composition.Hosting;
using System.Linq;
using System.Reflection;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using MeffedNavigationService.Infrastructure.DataModel;
using MeffedNavigationService.Infrastructure.Contract;

namespace MeffedNavigationService.ViewModel {
    public class MenuViewModel : ViewModel {
        IEnumerable<MenuDataItem> _items;
        [ImportMany]
        private IEnumerable<Lazy<IModule, IModuleMetadata>> _views;
        public MenuViewModel() {
            Items = null;
            this.InitializeCommands();
        }
        [ImportMany(typeof(IModuleListing))]
        private IEnumerable<Lazy<IModuleListing>> AddOnModuleListing { get; set; }
        public IEnumerable<MenuDataItem> Items {
            get { return _items; }
            private set { this.SetProperty<IEnumerable<MenuDataItem>>(ref this._items, value, "Items"); }
        }
        public IEnumerable<Lazy<IModule, IModuleMetadata>> Views {
            get { return this._views; }
        }
        public ICommand NavigationCommand { get; private set; }
        public void LoadState(object navigationParameter) {
            Items = new ObservableCollection<MenuDataItem>(SampleMenuDataSource.Instance.Items);
            // Load Module Listings.
            MeffedNavigationService.MEF.CompositionContainerHelper.ComposeParts(this, new DirectoryCatalog("."));
            foreach(Lazy<IModuleListing> addOnModule in AddOnModuleListing) {
                foreach(MenuDataItem menuItem in addOnModule.Value.MenuItems) {
                    ((ObservableCollection<MenuDataItem>)Items).Add(menuItem);
                }
            }
        }
        protected override void OnNavigatedTo(NavigationEventArgs e) {
            LoadState(e.Parameter);
        }
        private void InitializeCommands() {
            this.NavigationCommand = new DelegateCommand<MenuDataItem>(DoNavigation);
        }
        private void DoNavigation(MenuDataItem module) {
            NavigationService.Navigate(module.ModuleType.ToString());
        }
    }
}

File: VB/MeffedNavigationService/ViewModel/MenuViewModel.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Xpf.Mvvm
Imports DevExpress.Xpf.Mvvm.UI
Imports DevExpress.Xpf.WindowsUI.Navigation
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel.Composition
Imports System.ComponentModel.Composition.Hosting
Imports System.Linq
Imports System.Reflection
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Input
Imports MeffedNavigationService.Infrastructure.DataModel
Imports MeffedNavigationService.Infrastructure.Contract

Namespace MeffedNavigationService.ViewModel
	Public Class MenuViewModel
		Inherits ViewModel
		Private _items As IEnumerable(Of MenuDataItem)
		<ImportMany> _
		Private _views As IEnumerable(Of Lazy(Of IModule, IModuleMetadata))
		Public Sub New()
			Items = Nothing
			Me.InitializeCommands()
		End Sub
		Private privateAddOnModuleListing As IEnumerable(Of Lazy(Of IModuleListing))
		<ImportMany(GetType(IModuleListing))> _
		Private Property AddOnModuleListing() As IEnumerable(Of Lazy(Of IModuleListing))
			Get
				Return privateAddOnModuleListing
			End Get
			Set(ByVal value As IEnumerable(Of Lazy(Of IModuleListing)))
				privateAddOnModuleListing = value
			End Set
		End Property
		Public Property Items() As IEnumerable(Of MenuDataItem)
			Get
				Return _items
			End Get
			Private Set(ByVal value As IEnumerable(Of MenuDataItem))
				Me.SetProperty(Of IEnumerable(Of MenuDataItem))(Me._items, value, "Items")
			End Set
		End Property
		Public ReadOnly Property Views() As IEnumerable(Of Lazy(Of IModule, IModuleMetadata))
			Get
				Return Me._views
			End Get
		End Property
		Private privateNavigationCommand As ICommand
		Public Property NavigationCommand() As ICommand
			Get
				Return privateNavigationCommand
			End Get
			Private Set(ByVal value As ICommand)
				privateNavigationCommand = value
			End Set
		End Property
		Public Sub LoadState(ByVal navigationParameter As Object)
			Items = New ObservableCollection(Of MenuDataItem)(SampleMenuDataSource.Instance.Items)
			' Load Module Listings.
			MeffedNavigationService.MEF.CompositionContainerHelper.ComposeParts(Me, New DirectoryCatalog("."))
			For Each addOnModule As Lazy(Of IModuleListing) In AddOnModuleListing
				For Each menuItem As MenuDataItem In addOnModule.Value.MenuItems
					CType(Items, ObservableCollection(Of MenuDataItem)).Add(menuItem)
				Next menuItem
			Next addOnModule
		End Sub
		Protected Overrides Sub OnNavigatedTo(ByVal e As NavigationEventArgs)
			LoadState(e.Parameter)
		End Sub
		Private Sub InitializeCommands()
			Me.NavigationCommand = New DelegateCommand(Of MenuDataItem)(AddressOf DoNavigation)
		End Sub
		Private Sub DoNavigation(ByVal [module] As MenuDataItem)
			NavigationService.Navigate([module].ModuleType.ToString())
		End Sub
	End Class
End Namespace